man/org.freedesktop.login1.xml

```diff
@@ -1074,13 +1098,63 @@ node /org/freedesktop/login1/user/_1000 {
+      in <citerefentry><refentrytitle>logind.conf</refentrytitle><manvolnum>5</manvolnum></citerefentry>).
+      Callers of this method should first subscribe to the <function>PrepareForSecureLock()</function> signal.
+      Clients should use this mechanism to wipe any sensitive data out of memory (e.g. encryption keys, login
+      tokens), in preparation for a secure lock. They should also close open FDs, if possible, to give the kernel
+      a better chance at wiping the file contents out of memory. Clients should also use this mechanism to stop
+      interacting with system DBus services; a frozen client cannot read messages out of the queue, and if the queue
+      fills up the DBus broker might conclude that the client is unresponsive and disconnect it from the bus. Thus,
+      clients should unsubscribe from signals that they are watching and try to finish as many outstanding method
+      calls as possible before closing the FD.</para>
+
+      <para><function>InhibitAutoSecureLock()</function> temporarily inhibits automatic secure locking during
+      system suspend for this user. It returns a file descriptor that inhibits this functionality for as long
+      as it is open. As mentioned above, securely locking a user requires a separate lock screen running outside
+      of the user context, and will freeze the user session. Thus, securely locking a user is a trade-off: it
+      increases security, but prevents the lock screen from displaying any user content, including notifications,
+      media controls, contact information for incoming phone calls, and much more. A client may use this method to
+      implement more complicated automatic locking behavior for home directories, in order to solve some of these
+      UX issues. For instance, the client may choose to securely lock the user and switch to the secure lock screen
+      only after the device is inactive for over 24 hours. Note that this inhibitor does not prevent clients from
+      calling <function>SecureLock()</function> directly, and in fact clients will need to do so as part of their
+      custom behavior.</para>
+
+      <para><function>EnableSecureLock()</function> enables secure lock functionality for this user,
+      by setting up the "back-end" that will actually implement the locking. This should generally be called
+      by the service that is providing the user record to the system, like <filename>systemd-homed</filename>
+      or similar. When a secure lock should be activated, <filename>systemd-logind</filename> will emit the
+      <function>TriggerSecureLock()</function> signal. In response, the back-end should do whatever it needs
+      to securely lock the user, and then close the FD returned by this method to signal that it is done. The
+      back-end can also close the FD early to signal that secure locking has become unavailable for this user.
+      </para>
```

Comment:
i'd prefer if this would be done the other way round. i.e. homed should maybe just announce it's capability in some user record field it adds to the "state" section of the record or so.

and then maybe extend the userdbd varlink API with a lock function. if logind wants to lock homed it would asynchronously fire a varlink call to that.

Reply:
i mean, this is ultimately sort of a feature of the userdb record provider, and hence should probably be requested the way you usually talk to a userdb provider.

this would then mean homed never has to call into logind for this. Instead logind would do a varlink call into homed when it needs to via the generic userdb api.

Reply:
Overall I'm not against this, and your concerns make sense to me. However, so far the generic userdb API has been entirely read-only to the user state. Layering-wise it makes total sense, but API wise it doesn't IMO. Maybe putting it there is the wrong place.

How about we introduce a separate varlink API for the secure-lock-backend (with a signal method to lock a user), and put it somewhere that logind can consume it. For example: `/run/systemd/secure-lock-backend/$SERVICE` where `$SERVICE` corresponds to the service in the user record. So let's say logind wants to secure-lock a user managed by homed: it'll look up the service in the user record, find `io.systemd.Home`, then connect to `/run/systemd/secure-lock-backend/io.systemd.Home` and call `LockUser(username)`.

On the user-record side, we'd document that `status.state` can be `locked` so that clients like logind can rely on it.

Then logind will have everything it needs:
- Check whether or not secure lock is possible: check whether or not `/run/systemd/secure-lock-backend/$SERVICE` exists
- Check whether or not the user is already secure locked: check `status.state` in the user record for `locked`
- Activate the secure lock for the user: Call the private varlink API

Thoughts?

Reply:
people have asked for update apis in the userdb api since day one. I figure eventually we have to give them something, hence maybe sooner or later userdb should just have a generic api "UpdateUserRecord" or so. in other words: the fact that userdb stuff is read-only right now is not necessarily a design decision but just effect because noone added writability yet...

hence I am not sure it's really worth it trying to keep the secure lock stuff separate. 

that said, I am not opposed to doing this on a separate interface if you want, but please drop the "backend" string from the socket path. /run/systemd/secure-lock/<service> or /run/systemd/userdb-secure-lock/ sounds ok?

Reply:
> people have asked for update apis in the userdb api since day one

Eh I think this is a bad idea. Different backends will have wildly different semantics about how they'll be updated. Different clients will need to implement these different semantics anyways. Might as well communicate directly with the backend

In other words, I'm somewhat against this. IMO the userdb stuff should be read-only, and if you want to write you need to look at `service` and figure out if you know how to write for that `service` backend.

> i think it might make sense to add a field canSecureLock

100% makes sense :+1: